File: semester_2/homework_7/7.1(Stack, List)/7.1(Stack, List).Tests/StackTests.cs
namespace _7thHomework.Task1.Tests
{
    using System;
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    [TestClass]
    public class StackTests
    {
        private GenericStack<int> stack;

        [TestInitialize]
        public void Initialize()
        {
            stack = new GenericStack<int>();
        }

        [TestMethod]
        public void PushTest()
        {
            stack.Push(1);
            Assert.IsTrue(stack.IsBelong(1));
        }

        [TestMethod]
        public void PopTest()
        {
            stack.Push(1);
            Assert.AreEqual(1, stack.Pop());
        }

        [TestMethod]
        public void PeekTest()
        {
            stack.Push(1);
            Assert.AreEqual(1, stack.Peek());
        }

        [TestMethod]
        public void TwoElementsPopTest()
        {
            stack.Push(1);
            stack.Push(2);
            Assert.AreEqual(2, stack.Pop());
            Assert.AreEqual(1, stack.Pop());
        }

        [TestMethod]
        [ExpectedException(typeof(Exception))]
        public void PopFromEmptyStackTest()
        {
            stack.Pop();
        }
    }
}

File: semester_4/Work/Tests/StackTests.fs
namespace StackTests

open Microsoft.VisualStudio.TestTools.UnitTesting
open FsUnit.MsTest
open Work.Stack

[<TestClass>]
type StackTest() =

    [<TestMethod>]
    member this.``IsEmpty for empty stack should return True``() =
        let stack = new Stack<int>()
        stack.IsEmpty() |> should be True

    [<TestMethod>]
    member this.``IsEmpty for stack with elements should return False``() =
        let stack = new Stack<int>()
        stack.Push 14
        stack.IsEmpty() |> should be False

    [<TestMethod>]
    member this.``IsEmpty should return True``() =
        let stack = new Stack<int>()
        stack.Push 14
        stack.Pop() |> ignore
        stack.IsEmpty() |> should be True
    
    [<TestMethod>]
    member this.``Check Push and Pop``() =
        let stack = new Stack<int>()
        stack.Push 14
        stack.Push 15
        stack.Push 16
        stack.Pop() |> should equal 16
        stack.Pop() |> should equal 15
        stack.Pop() |> should equal 14

    [<TestMethod>]
    member this.``If Pop for empty stack, stack should throw exception``() =
        let stack = new Stack<int>()
        (fun () -> stack.Pop() |> ignore) |> should throw typeof<System.Exception>